### PR TITLE
[WIP] Character markers for away missions

### PR DESCRIPTION
**Pull request by unid15**
Character markers are invisible objects that can be placed in away missions and any maps loaded at runtime, under a mob. Then, observers and ghosts will be able to enter that mob

Here's what it roughly looks like (messages will be changed): 

There are two types of markers: team markers and individual markers. The difference between them is that individual markers each send a message to all observers, allowing observers to choose which mob to enter. Team markers with the same name form "teams", and each team sends a message to observers, allowing observers to enter a random mob from the team.

Syndicate-banned players can't use markers

Currently not compatible with vaults, because the message is only sent once, and tehre are usually no observers when vaults are spawned

Other changes:
* Added spawned_by_map_element(datum/map_element/ME, list/objects) proc, moved some stuff to it
* Turfs now hold a reference to their map element
* Added a new area type: /area/vault/automap. Using it in vaults removes the need of creating an entire different subtype of /area/vault for the map element (multiple instances of /area/vault/automap aren't linked together).

**TODO LIST**:

[V] Add better messages, find and fix bugs

  [V] Clear mob references 

[V] Admin logging

* Add an example map (like a trader shuttle)

* Add a way for ghosts to see all current vacations and enter them

### CODE REVIEW

**Comment by Shadowmech88** [comment]
You should add a version that admins can place and activate on the fly, to make getting players into mobs a bit more streamlined than saying "ahelp if you want me to manually edit your ckey into this mob I have" in deadchat.

File: code/game/objects/effects/mapping.dm [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	//O: observer that is volunteering
+	//body: either a mob or a list of mobs. If allow_choosing_
+	if(istype(body, /mob/living))
+		if(!isnull(body.client))
+			to_chat(O, "<span class='info'>The role has already been taken.</span>")
+	else
+		return
+
+	body.key = O.key
+	to_chat(body, "don't metagame, fucker") //todo
```
Comment (PJB3005) [inline code/game/objects/effects/mapping.dm]: Reminder.

File: code/game/objects/effects/mapping.dm [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+/obj/effect/mapping/playable_mob/proc/check_observer(var/mob/dead/observer/O)
+	if(O.has_enabled_antagHUD == 1 && config.antag_hud_restricted)
+		return 0
+	if(jobban_isbanned(O, "Syndicate")) // Antag-banned
+		return 0
+	if(!O.client)
+		return 0
+	if(((O.client.inactivity/10)/60) <= ALIEN_SELECT_AFK_BUFFER) // Filter AFK
```
Comment (PJB3005) [inline code/game/objects/effects/mapping.dm]: Pretty sure we have a proc for this.

**Comment by Skullyton** [comment]
Needs something to limit what a player can say when they're in control of a mob, maybe to their talk list, else you'll have shit like a hivebot running around shouting 'GAY NIGGERS EAT MY ANUS'

File: code/datums/map_elements.dm [hunk 3]
```diff
@@ -23,12 +23,8 @@ var/list/datum/map_element/map_elements = list()
 	if(!location && objects.len)
 		location = locate(/turf) in objects
 
-	//Build powernets
-	for(var/obj/structure/cable/C in objects)
```
Comment (ghost) [inline code/datums/map_elements.dm:27]: ???

File: vgstation13.dme [hunk 4]
```diff
@@ -2053,7 +2055,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_box.dm"
```
Comment (Kurfursten) [inline vgstation13.dme]: truly this will become the meme of the new era

Comment (Shadowmech88) [reply vgstation13.dme]: One of these days someone's not going to catch this and the players will have to endure an actual round on test_box.

File: code/datums/map_elements.dm (hunk 3 above)
Comment (JustSumBody) [reply code/datums/map_elements.dm:27]: It got moved to the `spawned_by_map_element` proc in cable.dm, which is called in the new lines here.

File: code/game/objects/effects/mapping.dm (hunk 1 above)
Comment (Intigracy) [reply code/game/objects/effects/mapping.dm]: :joy:

**Comment by icantthinkofanameritenow** [comment]
It would probably be a good idea to give players a way to use any spell-like abilities that the mob can use too. I remember once I was in the position of controlling a giant spider that couldn't even ventcrawl.

**Review by Intigracy — CHANGES_REQUESTED** [review]
pls fix test_box.dm include

**Comment by unid15** [comment]
@Skullyton that issue that will rarely occur. and when it will, there are admins to handle that

@Shadowmech88 I'll only include mapping ones for now

**Comment by unid15** [comment]
The functionality is ready, the only thing that is missing are example maps and some luxury features (like being able to see all current vacations)

**Review by Intigracy — APPROVED** [review]

**Comment by Probe1** [comment]
Please fix what conflicts you can this is really cool.

**Comment by unid15** [comment]
Will reopen later

**Pull request closed without merging**